core/ray.h

```diff
@@ -6,12 +6,19 @@
 
 #include <initializer_list>
 #include <iostream>
+#include <limits>
 
 namespace core {
 class Ray {
 public:
-  Ray(const Vec3 &origin = Vec3(0, 0, 0), const Vec3 &direction = Vec3(0, 0, 1),
-      float energy = 0);
+  // Returns Ray from |origin| and direction calculated from given spherical
+  // coordinates, where |psiRotation| represents rotation on z-axis and |phi|
+  // represents rotation on y-axis. Both |psi| and |phi| must be in radians.
+  static Ray makeRayFromSphericalCoords(const Vec3 &origin, const float psi,
```

Comment:
1. Do you think having two 3 latter parameters that have almost the same letters is a good idea?
2. You can also drop "const", not useful in case of scalars like this.
3. Make this function name shorter, think about call site when naming things. You will have `Ray::makeRayFromSphericalCoords`, which is a big long. How about just `fromSphericalCoords` so it reads nicely as `Ray::fromSphericalCoords(...)`